src/features/unified-share-modal/EmailForm.js

```diff
@@ -81,21 +82,37 @@ class EmailForm extends React.Component<Props, State> {
     } = React.createRef();
 
     handleContactAdd = (contacts: Array<Contact>) => {
-        const { onContactAdd, updateSelectedContacts } = this.props;
+        const { intl, contactLimit, selectedContacts, onContactAdd, updateSelectedContacts } = this.props;
 
-        updateSelectedContacts([...this.props.selectedContacts, ...contacts]);
+        updateSelectedContacts([...selectedContacts, ...contacts]);
+
+        let contactsFieldError = '';
+        if (contactLimit !== undefined && selectedContacts.length + contacts.length > contactLimit) {
+            contactsFieldError = intl.formatMessage(messages.contactsExceedLimitError, {
+                maxContacts: contactLimit,
+            });
+        }
+        this.setState({ contactsFieldError });
```

Comment:
It seems like you're doing this calculation and setting state 3 times in handleContactAdd, handleContactRemove, and handleSubmit.  Can we move it to componentDidUpdate?